pkg/proxy/nftables/proxier.go

```diff
@@ -1015,6 +1045,104 @@ func isAffinitySetName(set string) bool {
 	return strings.HasPrefix(set, servicePortEndpointAffinityNamePrefix)
 }
 
+// nftElementStorage is an internal representation of nftables map with single string value or set.
+// Currently only supports sets and maps with single string value.
+type nftElementStorage struct {
+	elements      map[string]string
+	leftoverKeys  sets.Set[string]
+	containerType string
+	containerName string
+}
+
+func (c *nftElementStorage) joinKey(k []string) string {
+	return strings.Join(k, " . ")
+}
+
+func (c *nftElementStorage) splitKey(k string) []string {
+	return strings.Split(k, " . ")
+}
+
+// newNFTElementStorage creates an empty nftElementStorage.
+// nftElementStorage.reset() must be called before the first usage.
+func newNFTElementStorage(containerType, containerName string) *nftElementStorage {
+	c := &nftElementStorage{
+		elements:      make(map[string]string),
+		leftoverKeys:  sets.New[string](),
+		containerType: containerType,
+		containerName: containerName,
+	}
+	return c
+}
+
+// reset clears the internal state and flushes the nftables map/set.
+func (c *nftElementStorage) reset(tx *knftables.Transaction) {
+	clear(c.elements)
+	if c.containerType == "set" {
+		tx.Flush(&knftables.Set{
+			Name: c.containerName,
+		})
+	} else {
+		tx.Flush(&knftables.Map{
+			Name: c.containerName,
+		})
+	}
+}
+
+// resetLeftoverKeys should be called before every transaction to ensure cleanupLeftoverKeys
+// will have correct data.
+func (c *nftElementStorage) resetLeftoverKeys() {
+	clear(c.leftoverKeys)
+	for key := range c.elements {
+		c.leftoverKeys.Insert(key)
+	}
+}
+
+// getValue converts element value to string.
+// It returns empty string for sets, and the first string for maps.
+// If a map value has 0 elements, it returns an empty string, but the following transaction will fail.
+func (c *nftElementStorage) getValue(v []string) string {
+	if c.containerType == "set" || len(v) == 0 {
+		return ""
+	} else {
+		return v[0]
+	}
+}
```

Comment:
I would drop this function and just do `newValue := c.joinKey(elem.Value)` in `ensureElem`.
If `elem.Value` is empty (as it always will be for sets), `strings.Join()` will just immediately return `""`, which is the same as what you do here. If `elem.Value` has length 1, then `strings.Join()` will just immediately return `elem.Value[0]`, which is the same as what you do here. And if, in the future, we start using maps with multi-valued values, then the code will still work rather than mysteriously failing.

(But now you need to rename `joinKey` because it's not just for keys...)

Reply:
right, that was an attempt to optimize joining strings considering we only ever use 1-string values for now, but now that I checked, strings.Join is already optimized for that (: